server/src/main/java/org/apache/iotdb/db/mpp/buffer/SinkHandle.java

```diff
@@ -0,0 +1,346 @@
+  private final LocalMemoryManager localMemoryManager;
+  private final ExecutorService executorService;
+  private final DataBlockService.Client client;
+  private final TsBlockSerde serde;
+  private final SinkHandleListener sinkHandleListener;
+
+  // TODO: a better data structure to hold tsblocks
+  private final List<TsBlock> bufferedTsBlocks = new LinkedList<>();
+
+  private volatile ListenableFuture<Void> blocked = immediateFuture(null);
+  private NewDataBlockEvent savedNewDataBlockEvent;
+  private int numOfBufferedTsBlocks = 0;
+  private long bufferRetainedSizeInBytes;
+  private boolean closed;
+  private boolean noMoreTsBlocks;
+  private Throwable throwable;
+
+  public SinkHandle(
+      String remoteHostname,
+      TFragmentInstanceId remoteFragmentInstanceId,
+      String remoteOperatorId,
+      TFragmentInstanceId localFragmentInstanceId,
+      LocalMemoryManager localMemoryManager,
+      ExecutorService executorService,
+      DataBlockService.Client client,
+      TsBlockSerde serde,
+      SinkHandleListener sinkHandleListener) {
+    this.remoteHostname = Validate.notNull(remoteHostname);
+    this.remoteFragmentInstanceId = Validate.notNull(remoteFragmentInstanceId);
+    this.remoteOperatorId = Validate.notNull(remoteOperatorId);
+    this.localFragmentInstanceId = Validate.notNull(localFragmentInstanceId);
+    this.localMemoryManager = Validate.notNull(localMemoryManager);
+    this.executorService = Validate.notNull(executorService);
+    this.client = Validate.notNull(client);
+    this.serde = Validate.notNull(serde);
+    this.sinkHandleListener = Validate.notNull(sinkHandleListener);
+  }
+
+  @Override
+  public ListenableFuture<Void> isFull() {
+    if (closed) {
+      throw new IllegalStateException("Sink handle is closed.");
+    }
+    return nonCancellationPropagating(blocked);
+  }
+
+  private void submitSendNewDataBlockEventTask(int startSequenceId, List<Long> blockSizes) {
+    executorService.submit(new SendNewDataBlockEventTask(startSequenceId, blockSizes));
+  }
+
+  @Override
+  public void send(List<TsBlock> tsBlocks) throws IOException {
+    Validate.notNull(tsBlocks, "tsBlocks is null");
+    if (throwable != null) {
+      throw new IOException(throwable);
+    }
+    if (closed) {
+      throw new IllegalStateException("Sink handle is closed.");
+    }
+    if (!blocked.isDone()) {
+      throw new IllegalStateException("Sink handle is blocked.");
+    }
+    if (noMoreTsBlocks) {
+      return;
+    }
+
+    long retainedSizeInBytes = 0L;
+    for (TsBlock tsBlock : tsBlocks) {
+      retainedSizeInBytes += tsBlock.getRetainedSizeInBytes();
+    }
+    int currentEndSequenceId;
+    List<Long> tsBlockSizes = new ArrayList<>();
+    synchronized (this) {
+      currentEndSequenceId = bufferedTsBlocks.size();
+      blocked =
+          localMemoryManager
+              .getQueryPool()
+              .reserve(localFragmentInstanceId.getQueryId(), retainedSizeInBytes);
```

Comment:
Yes. Since SinkHandle provides `isFull` for callers to decide if they are allowed to call `send`, it is safe to assume that the listenable future `blocked` is completed.